src/widget/transfer_function.ts

```diff
@@ -0,0 +1,1625 @@
+          if (firstPointIndexInWindow > 0) {
+            const pointBeforeWindow =
+              normalizedControlPoints[firstPointIndexInWindow - 1];
+            const interpFactor = computeInvlerp(
+              [pointBeforeWindow.input, firstPointInWindow.input],
+              -1,
+            );
+            const lineStartY = computeLerp(
+              [pointBeforeWindow.output, firstPointInWindow.output],
+              DataType.FLOAT32,
+              interpFactor,
+            ) as number;
+            lineFromLeftEdge = vec4.fromValues(
+              -1,
+              lineStartY,
+              firstPointInWindow.input,
+              firstPointInWindow.output,
+            );
+          }
+          // If the first point in the window is the leftmost point, draw a 0 line up to the point
+          else {
+            lineFromLeftEdge = vec4.fromValues(
+              firstPointInWindow.input,
+              -1,
+              firstPointInWindow.input,
+              firstPointInWindow.output,
+            );
+          }
+          numLines += 1;
+        }
+
+        // Need to draw a line from the last control point in the window to the right edge
+        const lastPointInWindow =
+          normalizedControlPoints[lastPointIndexInWindow!];
+        if (lastPointInWindow.input < 1) {
+          // If there is a value to the right, draw a line from the last point in the window to the point outside the window
+          if (lastPointIndexInWindow! < controlPoints.length - 1) {
+            const pointAfterWindow =
+              normalizedControlPoints[lastPointIndexInWindow! + 1];
+            const interpFactor = computeInvlerp(
+              [lastPointInWindow.input, pointAfterWindow.input],
+              1,
+            );
+            const lineEndY = computeLerp(
+              [lastPointInWindow.output, pointAfterWindow.output],
+              DataType.FLOAT32,
+              interpFactor,
+            ) as number;
+            lineToRightEdge = vec4.fromValues(
+              lastPointInWindow.input,
+              lastPointInWindow.output,
+              1,
+              lineEndY,
+            );
+          }
+          // If the last point in the window is the rightmost point, draw a line from the point to 1
+          else {
+            lineToRightEdge = vec4.fromValues(
+              lastPointInWindow.input,
+              lastPointInWindow.output,
+              1,
+              lastPointInWindow.output,
+            );
+          }
+          numLines += 1;
+        }
+      }
+    }
+
+    const linePositionArray = new Float32Array(
+      numLines * POSITION_VALUES_PER_LINE * VERTICES_PER_LINE,
+    );
+
+    if (lineFromLeftEdge !== null) {
+      positionArrayIndex = addLine(
+        linePositionArray,
+        positionArrayIndex,
+        lineFromLeftEdge,
+      );
+    }
+
+    // Update points and draw lines between control points
+    for (let i = 0; i < controlPoints.length; ++i) {
```

Comment:
This currently duplicates the code above at the ends. Needs check for input and output between -1 and 1